Domain/SecurityIdentityRetrievalStrategy.php

```diff
@@ -39,37 +39,52 @@ public function __construct(RoleHierarchyInterface $roleHierarchy, Authenticatio
 
     /**
      * {@inheritdoc}
+     *
+     * @return RoleSecurityIdentity[]
      */
-    public function getSecurityIdentities(TokenInterface $token)
+    public function getSecurityIdentities(?TokenInterface $token)
     {
         $sids = [];
 
         // add user security identity
-        if (!$token instanceof AnonymousToken) {
+        if (!$token instanceof AnonymousToken && null !== $token) {
             try {
                 $sids[] = UserSecurityIdentity::fromToken($token);
             } catch (\InvalidArgumentException $e) {
                 // ignore, user has no user security identity
             }
         }
 
-        // add all reachable roles
-        foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
-            $sids[] = new RoleSecurityIdentity($role);
+        if (null !== $token) {
+            // add all reachable roles
+            foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
+                $sids[] = new RoleSecurityIdentity($role);
+            }
         }
 
+        $anonymousRole = \defined('\Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter::PUBLIC_ACCESS') ? AuthenticatedVoter::PUBLIC_ACCESS : AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY;
         // add built-in special roles
         if ($this->authenticationTrustResolver->isFullFledged($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_FULLY);
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         } elseif ($this->authenticationTrustResolver->isRememberMe($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
-        } elseif ($this->authenticationTrustResolver->isAnonymous($token)) {
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
+        } elseif ($this->isNotAuthenticated($token)) {
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         }
 
         return $sids;
     }
+
+    private function isNotAuthenticated(?TokenInterface $token): bool
+    {
+        if (\defined('\Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter::PUBLIC_ACCESS')) {
+            return !$this->authenticationTrustResolver->isAuthenticated($token);
+        }
+
+        /** @psalm-suppress UndefinedInterfaceMethod */
```

Comment:
```suggestion
        // @psalm-suppress UndefinedInterfaceMethod
```

Can you check like this? The other option is to add the suppress directly on the method to avoid php cs fixer complains

Reply:
Both above and behind the line doesn't seem to work in pslam 🤔 

```php
return $this->authenticationTrustResolver->isAnonymous($token); // @psalm-suppress UndefinedInterfaceMethod
```